File: Algos/mergeSort.js
function mergeSort(array) {
  // base case, return when there is at most 1 element in the array since subarray si sorted
  if (array.length <= 1) {
    return array
  }
  const middleIndex = Math.floor(array.length / 2);
  return merge(
    mergeSort(array.slice(0, middleIndex)),
    mergeSort(array.slice(middleIndex))
  );
}

function merge(left, right) {
  let merged = [];
  let i = 0, j = 0;
  while (i < left.length && j < right.length) {
    if (left[i] < right[j]) {
      merged.push(left[i]);
      i++;
    } else {
      merged.push(right[j]);
      j++;
    }
  }
  return merged.concat(left.slice(i)).concat(right.slice(j));
}

console.log(mergeSort([3, 2, 1, 13, 8, 5, 0, 1]));
console.log(mergeSort([105, 79, 100, 110]));

File: Algos/merge_sort.py
def mergeSort(array):
  if len(array) <= 1:
    return array
  middle = len(array) // 2
  return merge(
    mergeSort(array[:middle]),
    mergeSort(array[middle:])
    )

def merge(left, right):
  if not left:
    return right
  if not right:
    return left
  if left[0] < right[0]:
    return [left[0]] + merge(left[1:], right)
  return [right[0]] + merge(left, right[1:])

print(mergeSort([7, 6, 5, 4, 3, 2, 1]))